### PR TITLE
Add support for wrapped lists to push and pop items

### PR DESCRIPTION
**Pull request by ejsmith**
Test that is showing that trying to push to an interop list does not work. Should this work? Is there some rule that I'm not aware of that is trying to keep interop to read only operations?

### CODE REVIEW

**Comment by lahma** [comment]
Jint doesn't automatically attach array prototype when wrapping objects ([same as with STJ](https://github.com/sebastienros/jint/blob/23a2a2587750a98cd55e4697e4da8270bcf460db/Jint.Tests.PublicInterface/InteropTests.SystemTextJson.cs)). Here's a good example which it would a bad decision by default. JS expects that you can index outside of collection size to create a new entry. Probably a custom `ObjectWrapper` implementation catching the outside-of-bounds indexer usage and calling `Add` instead for target might work.

**Comment by ejsmith** [comment]
Hmm... I guess that is kind of confusing that it implements things like `indexOf` but not `push`. I will experiment with a `IObjectConverter` to see if I can make it support `push`.

**Comment by lahma** [comment]
IndexOf is  a List method, it doesn't need JS array prototype

**Comment by ejsmith** [comment]
Hmm... yeah, I guess a lot of the JS Array methods line up with the list methods. I tried what you said and it didn't work either.

```csharp
        [Fact]
        public void ArrayPrototypePushWithInteropList()
        {
            var engine = new Jint.Engine(options =>
            {
                // make List behave like JS array
                options.Interop.WrapObjectHandler = static (e, target, type) =>
                {
                    if (target is IList)
                    {
                        var wrapped = ObjectWrapper.Create(e, target);
                        wrapped.Prototype = e.Intrinsics.Array.PrototypeObject;
                        return wrapped;
                    }

                    return ObjectWrapper.Create(e, target);
                };
            });

            engine.SetValue("list", new List<string> { "A", "B", "C" });

            engine.Evaluate("list.Add('D')");
            Assert.Equal(3, engine.Evaluate("list.LastIndexOf('D')"));
        }
```

**Comment by lahma** [comment]
You probably need to implement something custom deriving from ObjectWrapper which would intercept problematic calls.

**Comment by ejsmith** [comment]
`ObjectWrapper` is sealed.

**Comment by ejsmith** [comment]
This seems like something that should work out of the box. Do you agree with that or is against the direction you want to go with the project?

I am going to investigate more and try and figure out the correct spot to implement this. The code is quite complex so trying to figure out the correct way to implement it will probably take some time. I'd be willing to send some sponsor money your way to either give me a tour of the code and point me in the right direction or if you just want to implement yourself.

**Comment by ejsmith** [comment]
Kept digging and found a way that I think might be acceptable to get `push` and `pop` on wrapped `IList` objects.

File: Jint.Tests/Runtime/InteropTests.cs [hunk 1]
```diff
@@ -2913,6 +2913,76 @@ public void ArrayPrototypeIndexOfWithInteropList()
...
+            Assert.Equal('B', engine.Evaluate("list.find((x) => x === 'B')"));
+        }
+
+        [Fact]
+        public void ArrayPrototypePushWithInteropList()
+        {
+            var engine = new Jint.Engine(options =>
+            {
+                // make List behave like JS array
+                options.Interop.WrapObjectHandler = static (e, target, type) =>
```
Comment (lahma) [inline Jint.Tests/Runtime/InteropTests.cs:2934]: with latest main this is probably no longer needed

File: Jint.Tests/Runtime/InteropTests.cs [hunk 2]
```diff
@@ -2913,6 +2913,76 @@ public void ArrayPrototypeIndexOfWithInteropList()
...
+            Assert.Equal(3, engine.Evaluate("list.lastIndexOf('D')"));
+        }
+
+        [Fact]
+        public void ArrayPrototypePopWithInteropList()
+        {
+            var engine = new Jint.Engine(options =>
+            {
+                // make List behave like JS array
+                options.Interop.WrapObjectHandler = static (e, target, type) =>
```
Comment (lahma) [inline Jint.Tests/Runtime/InteropTests.cs:2963]: probably not needed

File: Jint/Native/Array/ArrayOperations.cs [hunk 3]
```diff
@@ -493,7 +493,31 @@ public IndexWrappedOperations(ObjectWrapper wrapper)
 
             public override ulong GetLongLength() => GetLength();
 
-            public override void SetLength(ulong length) => throw new NotSupportedException();
+            public override void SetLength(ulong length)
+            {
+                if (_list == null)
+                    throw new NotSupportedException();
```
Comment (lahma) [inline Jint/Native/Array/ArrayOperations.cs:499]: we always use braces

File: Jint/Native/Array/ArrayOperations.cs [hunk 4]
```diff
@@ -493,7 +493,31 @@ public IndexWrappedOperations(ObjectWrapper wrapper)
...
+                    // shrink list to fit
+                    for (var i = _list.Count - 1; i >= (int) length; i--)
+                    {
+                        _list.RemoveAt(i);
+                    }
+
+                    return;
+                }
+                
+                if (length >= (ulong) _list.Count)
```
Comment (lahma) [inline Jint/Native/Array/ArrayOperations.cs:512]: for loop should cover this if already

File: Jint/Native/Array/ArrayOperations.cs [hunk 5]
```diff
@@ -493,7 +493,31 @@ public IndexWrappedOperations(ObjectWrapper wrapper)
 
             public override ulong GetLongLength() => GetLength();
 
-            public override void SetLength(ulong length) => throw new NotSupportedException();
+            public override void SetLength(ulong length)
+            {
+                if (_list == null)
+                    throw new NotSupportedException();
+                
+                if (length < (ulong) _list.Count)
```
Comment (lahma) [inline Jint/Native/Array/ArrayOperations.cs:501]: could probably be just while loop, next for loop for addition will be skipped

File: Jint/Native/Array/ArrayOperations.cs [hunk 6]
```diff
@@ -493,7 +493,31 @@ public IndexWrappedOperations(ObjectWrapper wrapper)
...
+
+                    return;
+                }
+                
+                if (length >= (ulong) _list.Count)
+                {
+                    // expand list to fit
+                    for (var i = _list.Count; i < (int) length; i++)
+                    {
+                        _list.Add(JsValue.Undefined.ToObject());
```
Comment (lahma) [inline Jint/Native/Array/ArrayOperations.cs:517]: would be clearer just to call with `null`

File: Jint/Native/Object/ObjectInstance.cs [hunk 7]
```diff
@@ -702,6 +703,11 @@ internal bool DeletePropertyOrThrow(JsValue property)
         /// </summary>
         public virtual bool Delete(JsValue property)
         {
+            if (this is ObjectWrapper objectWrapper && objectWrapper.Target is IList && property.IsNumber())
```
Comment (lahma) [inline Jint/Native/Object/ObjectInstance.cs:706]: should override in ObjectWrapper instead of breaking the whole idea of OOP...

**Comment by ejsmith** [comment]
Updated. Thanks for the feedback!

**Comment by ejsmith** [comment]
> Thank you!

No, thank you!

**Pull request merged**